libvirt/resource_libvirt_domain.go

```diff
@@ -268,9 +277,17 @@ func resourceLibvirtDomainCreate(d *schema.ResourceData, meta interface{}) error
 		}
 	}
 
-	domainDef.OS.Type.Arch = d.Get("arch").(string)
-	domainDef.OS.Type.Machine = d.Get("machine").(string)
-	
+	if guestarch, ok := d.GetOk("arch"); ok {
+		domainDef.OS.Type.Arch = guestarch.(string)
+	}
+	if machine, ok := d.GetOk("machine"); ok {
+
+		domainDef.OS.Type.Machine = machine.(string)
+	}
+	if emulator, ok := d.GetOk("emulator"); ok {
+		domainDef.Devices.Emulator = emulator.(string)
+	}
+
```

Comment:
You can skip the whole `GetOk(...)` construct, and just use `Get(...)`.

Reply:
Wouldn't that overwrite the defaults with blanks if there wasn't something set in d ?

We need to keep the defaults from the connection otherwise it will see changes on creation.

Reply:
Not according to the [documentation](https://godoc.org/github.com/hashicorp/terraform/helper/schema#ResourceData.Get):

> If the key does exist in the schema but doesn't exist in the configuration, then the default value for that type will be returned. For strings, this is "", for numbers it is 0, etc.

And since we are setting a `Default` value in the [Schema](https://godoc.org/github.com/hashicorp/terraform/helper/schema#Schema), it won't be blank.

Reply:
But isn't 

` domainDef.Devices.Emulator = Get("emulator").(string)`

effectively going to call .(string) on a nil value (default return from Get for interface{}) which will replace with a blank ?

Sorry if I'm just wrong here, I'm pretty new to go.

Reply:
No, `d.Get("emulator")` will not return nil, I'm sure. It'll be either blank (`""`) or a specified default value (our case). Therefore, `.(string)` will not be called on a `nil` value. If it were `nil`, it would panic (which it doesn't) with:
```
panic: interface conversion: interface {} is nil, not string
```

Reply:
@monstermunchkin Changes as suggested and pushed